File: app/src/main/java/com/lostincontext/mainscreen/MainScreenFragment.java
package com.lostincontext.mainscreen;

import android.app.PendingIntent;
import android.content.Intent;
import android.content.res.Resources;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.lostincontext.R;
import com.lostincontext.commons.list.SpacesItemDecoration;
import com.lostincontext.data.rules.Rule;
import com.lostincontext.databinding.MainScreenFragmentBinding;
import com.lostincontext.ruledetails.RuleDetailsActivity;
import com.lostincontext.rulescreation.RulesCreationActivity;
import com.lostincontext.that.ThatService;

import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;


public class MainScreenFragment extends Fragment implements MainScreenContract.View {

    private MainScreenContract.Presenter presenter;
    private MainScreenAdapter adapter;


    public static MainScreenFragment newInstance() {
        return new MainScreenFragment();
    }


    public MainScreenFragment() { }


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        MainScreenFragmentBinding binding = DataBindingUtil.inflate(inflater, R.layout.main_screen_fragment, container, false);
        binding.setPresenter(presenter);


        RecyclerView recyclerView = binding.recyclerView;
        Resources resources = getResources();
        final int span = resources.getInteger(R.integer.list_span);
        GridLayoutManager layoutManager = new GridLayoutManager(getContext(), span);
        recyclerView.setLayoutManager(layoutManager);
        int space = resources.getDimensionPixelSize(R.dimen.grid_spacing);
        recyclerView.addItemDecoration(new SpacesItemDecoration(space, span));

        adapter = new MainScreenAdapter(presenter);

        layoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override public int getSpanSize(int position) {
                switch (adapter.getCurrentState()) {
                    case LOADING:
                    case ERROR:
                    case EMPTY:
                        return span;

                    case CONTENT:
                        return 1;

                    default:
                        throw new RuntimeException("invalid state");
                }
            }
        });
        recyclerView.setAdapter(adapter);
        return binding.getRoot();
    }

    @Override public void onResume() {
        super.onResume();
        presenter.start();
    }

    @Override public void setPresenter(MainScreenContract.Presenter presenter) {
        this.presenter = checkNotNull(presenter);
    }

    public PendingIntent getPendingIntent() {
        Intent intent = new Intent(this.getContext(), ThatService.class);
        return PendingIntent.getService(this.getContext().getApplicationContext(),
                                        0,
                                        intent,
                                        0);
    }


    @Override public void openRuleCreationScreen() {
        Intent intent = new Intent(this.getContext(), RuleDetailsActivity.class);
        startActivity(intent);
    }

    @Override public void setRules(List<Rule> rules) {
        adapter.setRules(rules);
    }

}

File: app/src/main/java/com/lostincontext/mainscreen/MainScreenFragment.kt
package com.lostincontext.mainscreen

import android.app.Activity
import android.app.PendingIntent
import android.content.Intent
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.lostincontext.R
import com.lostincontext.application.LostApplication
import com.lostincontext.awareness.AwarenessModule
import com.lostincontext.commons.BaseActivity
import com.lostincontext.commons.list.SpacesItemDecoration
import com.lostincontext.commons.list.StatefulAdapter.ContentState
import com.lostincontext.data.playlist.Playlist
import com.lostincontext.data.rulesV2.Rule
import com.lostincontext.databinding.MainScreenFragmentBinding
import com.lostincontext.playlists.PlaylistsContract
import com.lostincontext.rulecreate.RuleCreateActivity
import com.lostincontext.rulecreate.RuleCreateContract.EXTRA_RULE
import com.lostincontext.that.ThatService
import com.lostincontext.utils.logD
import javax.inject.Inject


class MainScreenFragment : Fragment(), MainScreenContract.View {

    @Inject internal lateinit var presenter: MainScreenPresenter

    private lateinit var adapter: MainScreenAdapter


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Create the presenter
        DaggerMainScreenComponent.builder()
                .mainScreenPresenterModule(MainScreenPresenterModule(this))
                .applicationComponent((activity.application as LostApplication).appComponent)
                .awarenessModule(AwarenessModule(activity as BaseActivity))
                .build()
                .inject(this)

    }


    override fun onCreateView(inflater: LayoutInflater?,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val binding = DataBindingUtil.inflate<MainScreenFragmentBinding>(inflater!!,
                                                                         R.layout.main_screen_fragment,
                                                                         container,
                                                                         false)
        binding.presenter = presenter


        val recyclerView = binding.recyclerView
        val resources = resources
        val span = resources.getInteger(R.integer.list_span)
        val layoutManager = GridLayoutManager(context, span)
        recyclerView.layoutManager = layoutManager
        val space = resources.getDimensionPixelSize(R.dimen.grid_spacing)
        recyclerView.addItemDecoration(SpacesItemDecoration(space, span))

        adapter = MainScreenAdapter(presenter)

        layoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                when (adapter.currentState) {
                    ContentState.LOADING,
                    ContentState.ERROR,
                    ContentState.EMPTY -> return span

                    ContentState.CONTENT -> return 1

                    else -> throw RuntimeException("invalid state")
                }
            }
        }
        recyclerView.adapter = adapter
        recyclerView.setHasFixedSize(true)
        return binding.root
    }

    override fun onResume() {
        super.onResume()
        presenter.start()
    }

    override fun getPendingIntent(playlist: Playlist): PendingIntent {
        val intent = Intent(this.context, ThatService::class.java)
        return PendingIntent.getService(this.context.applicationContext,
                                        0,
                                        intent,
                                        0)
    }


    override fun openRuleCreationScreen() {
        val intent = Intent(this.context, RuleCreateActivity::class.java)
        startActivityForResult(intent, CREATE_RULE_CODE)
    }

    override fun setRules(rules: List<Rule>) {
        adapter.setRules(rules)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        logD(TAG) { "onActivityResult : requestCode $requestCode, resultCode $resultCode, intent $data " }
        super.onActivityResult(requestCode, resultCode, data)
        if (data == null) return
        when (requestCode) {
            CREATE_RULE_CODE -> {
                if (resultCode == Activity.RESULT_OK) {
                    val rule = data.getParcelableExtra<Rule>(EXTRA_RULE)
                    presenter.onRuleInput(rule)
                }


            }
        }

    }

    override fun getPendingIntentFor(playlist: Playlist): PendingIntent {
        val intent = Intent(this.context, ThatService::class.java)
        return PendingIntent.getService(this.context.applicationContext,
                                        0,
                                        intent,
                                        0)
    }


    companion object {
        private val TAG: String = MainScreenFragment::class.java.simpleName
        fun newInstance(): MainScreenFragment = MainScreenFragment()
    }

}